doc/manual/src/contributing/testing.md

```diff
@@ -2,10 +2,56 @@
 
 ## Unit-tests
 
-The unit-tests for each Nix library (`libexpr`, `libstore`, etc..) are defined
-under `src/{library_name}/tests` using the
-[googletest](https://google.github.io/googletest/) and
-[rapidcheck](https://github.com/emil-e/rapidcheck) frameworks.
+The unit-tests for defined using the [googletest] and [rapidcheck] frameworks.
+
+[googletest]: https://google.github.io/googletest/
+[rapidcheck]: https://github.com/emil-e/rapidcheck
+
+### Source and header layout
+
+> An example for `libnixexpr`, demonstrating much of what is described below
+>
+> ```
+> src
+> ├── libexpr
+> │   ├── value/context.hh
+> │   ├── value/context.cc
+> │   │
+> │   …
+> ├── libexpr-tests
+> │   ├── tests/
+> │   │   ├── value/context.hh
+> │   │   │
+> │   │   …
+> │   ├── value/context.cc
+> │   │
+> │   …
+> …
+> ```
+
+The tests for each Nix library (`libnixexpr`, `libnixstore`, etc..) live inside a directory `src/${library_name_without-nix}-tests` next to the directory for the library (`src/${library_name_without-nix}`).
+Given a `foo/bar.{hh,cc}` interface (header) and implementation pair in the original library, we write tests for it in `foo/bar.cc`, and (possibly) declare additional interfaces in `tests/foo/bar.cc` in the test directory.
```

Comment:
Rather than the arbitrary `foo` and `bar` here, could you maybe reword this so it picks up the naming in the tree above? That would make it much more concrete, and much easier for me to understand.

Specifically, why is there a `libexpr-tests/value/context.cc` (I imagine this is just the tests for `libexpr/value/context.cc`) as well as a `libexpr-tests/tests/value/context.hh` (what's the difference between this and `libexpr/value/context.hh`)?

Reply:
the fix here is to prevent name collisions in the first place by separating out the directories, its less a case of concerning the contents of the files.

Reply:
```suggestion
Given a interface (header) and implementation pair in the original library, say, `value/context.{hh,cc}`, we write tests for it in `value/context.cc`, and (possibly) declare additional interfaces for testing purposes in `test/value/context.hh`.
(Note only the extra header is nested inside the `test` directory. This is explained further below.) 
```
How's this?

Reply:
Yes, that is much easier for me to wrap my head around! Thanks!